babel/numbers.py

```diff
@@ -676,17 +676,36 @@ def parse_decimal(string, locale=LC_NUMERIC):
         ...
     NumberFormatError: '2,109,998' is not a valid decimal number
 
+    If `strict` is set to `True` and the given string contains a number
+    formatted in an irregular way, an exception is raised:
+
+    >>> parse_decimal('30.00', locale='de', strict=True)
+    Traceback (most recent call last):
+        ...
+    NumberFormatError: '30.00' is not a properly formatted decimal number
+
     :param string: the string to parse
     :param locale: the `Locale` object or locale identifier
+    :param strict: controls whether numbers formatted in a weird way are
+                   accepted or rejected
     :raise NumberFormatError: if the string can not be converted to a
                               decimal number
     """
     locale = Locale.parse(locale)
+    group_symbol = get_group_symbol(locale)
+    decimal_symbol = get_decimal_symbol(locale)
     try:
-        return decimal.Decimal(string.replace(get_group_symbol(locale), '')
-                               .replace(get_decimal_symbol(locale), '.'))
+        r = decimal.Decimal(string.replace(group_symbol, '')
+                                  .replace(decimal_symbol, '.'))
     except decimal.InvalidOperation:
         raise NumberFormatError('%r is not a valid decimal number' % string)
+    if strict and group_symbol in string:
+        proper = format_decimal(r, locale=locale, decimal_quantization=False)
+        if string != proper and string.rstrip('0') != (proper + decimal_symbol):
+            raise NumberFormatError(
+                "%r is not a properly formatted decimal number" % string
```

Comment:
It would be super nice to have the expected value in the error here.

Maybe even

```python
class MisparsedNumberError(NumberFormatError):
  def __init__(self, message, original, parsed, expected):
    super(MisparsedNumberError, self).__init__(message)
    self.original = original
    self.parsed = parsed
    self.expected = expected

# ...

raise MisparsedNumberError(
  "%r is not properly formatted; parsed as %r, expected %r" % (string, r, proper),
  original=string,
  parsed=r,
  expected=proper,
)

```

so users could have more visibility into what went wrong.